gmcs/customize.py

```diff
@@ -4,46 +4,40 @@
 ######################################################################
 # imports
 
-import os
+import codecs
 import datetime
+import os
 import shutil
-from . import tdl
-#import tarfile
-#import gzip
-#import zipfile
 import sys
-#import re
-import codecs
+
 from subprocess import call
 
 from delphin import tsdb
 
+from gmcs import tdl
+
 from gmcs.choices import ChoicesFile
-#from gmcs.utils import TDLencode
-#from gmcs.utils import get_name
 from gmcs.utils import format_comment_block
 
-#from gmcs.lib import TDLHierarchy
-
-from gmcs.linglib import morphotactics
-from gmcs.linglib import information_structure
+from gmcs.linglib import adnominal_possession
+from gmcs.linglib import agreement_features
 from gmcs.linglib import argument_optionality
-from gmcs.linglib import direct_inverse
 from gmcs.linglib import case
-from gmcs.linglib import word_order
+from gmcs.linglib import clausalcomps
+from gmcs.linglib import clausalmods
+from gmcs.linglib import coordination
+from gmcs.linglib import direct_inverse
 from gmcs.linglib import features
+from gmcs.linglib import information_structure
 from gmcs.linglib import lexical_items
-from gmcs.linglib import agreement_features
-from gmcs.linglib import adnominal_possession
-from gmcs.linglib import verbal_features
+from gmcs.linglib import morphotactics
 from gmcs.linglib import negation
-from gmcs.linglib import coordination
-from gmcs.linglib import yes_no_questions
+from gmcs.linglib import nominalized_clauses
 from gmcs.linglib import toolboximport
 from gmcs.linglib import valence_change
-from gmcs.linglib import clausalmods
-from gmcs.linglib import nominalized_clauses
-from gmcs.linglib import clausalcomps
+from gmcs.linglib import verbal_features
+from gmcs.linglib import word_order
+from gmcs.linglib import yes_no_questions
```

Comment:
Interesting issue regarding circular imports. As you saw, I did run the regression tests (several times now 🙂), and they have all passed. I just tried the (sparse) unit tests and they have the same state as trunk. The example you raise shouldn't lead to a circular import, though, but is rather just a mutual dependency?

The goal of sorting imports is slightly more utilitarian than simply visually appealing. [reorder-python-imports](https://pypi.org/project/reorder-python-imports/) puts it pretty simply:
> The style chosen by reorder-python-imports has a single aim: reduce merge conflicts.

I suppose this isn't a huge issue for us, but in general having a consistent and easily applicable ordering reduces clutter and churn.

Reply:
> The example you raise shouldn't lead to a circular import, though, but is rather just a mutual dependency?

Yes, you're right. Maybe I shouldn't do PR reviews when it's late :P Nevertheless, there is at least one circular dependency between `gmcs/linglib/valence_change.py` and `gmcs/linglib/morphotactics.py`:

```console
$ grep import * | grep linglib
[...]
morphotactics.py:    from gmcs.linglib.valence_change import lexrule_name, added_argnum_for_vchop
valence_change.py:#from gmcs.linglib.morphotactics import all_position_classes
valence_change.py:    from gmcs.linglib.morphotactics import all_position_classes
[...]
```

It looks like we're getting around it by putting the import inside a function (indicated by the indentation here).

In general though I agree with your motivation for sorting the imports. So maybe your sorting is fine for now? It would be nice to eventually have a more principled way of avoiding potential import issues.

Reply:
I think in general, much of the importing that happens between libraries should be consolidated to constants or utils files.

This is also good for semantic reasons. e.g. the core basic lexicon types and basic morphology types are almost always defined inline, and it’s hard to tell if seeing one that is different (e.g. missing copulas) is different because of an oversight or an intentional decision.

Reply:
> I think in general, much of the importing that happens between libraries should be consolidated to constants or utils files.

Yes, and that sounds like a separate PR. Since nothing broke with the import sorting, I'll resolve this point.